applications/IgaApplication/custom_utilities/node_curve_geometry_3d.h

```diff
@@ -0,0 +1,206 @@
+     * 
+     * \param Index Index of the node
+     * 
+     * @return Location of the Kratos node at the given index.
+     */
+    VectorType Pole(
+        const int Index) const override
+    {
+        const auto& node = *Node(Index);
+ 
+        VectorType pole;
+        pole[0] = node[0];
+        pole[1] = node[1];
+        pole[2] = node[2];
+ 
+        return pole;
+    }
+
+    /** Sets the location of the Kratos node at a given index.
+     * 
+     * \param Index Index of the node
+     * \param Value New location of the Kratos node
+     */
+    void SetPole(
+        const int Index,
+        const VectorType& Value) override
+    {
+        auto& node = *Node(Index);
+ 
+        node[0] = Value[0];
+        node[1] = Value[1];
+        node[2] = Value[2];
+    }
+
+    /** Gets a value indicating whether or not the NURBS curve is rational.
+     * 
+     * @return True whether the curve is rational, otherwise false.
+     */
+    bool IsRational() const override
+    {
+        return true;
+    }
+
+    /** Gets the weight of the Kratos node at a given index.
+     * 
+     * \param Index Index of the node
+     * 
+     * @return Weight of the Kratos node at the given index.
+     */
+    ScalarType Weight(
+        const int Index) const override
+    {
+        const auto& node = *Node(Index);
+ 
+        return node.GetValue(Kratos::NURBS_CONTROLPOINT_WEIGHT);
+    }
+
+    /** Sets the weight of the Kratos node at a given index.
+     * 
+     * \param Index Index of the node
+     * \param Value New weight of the Kratos node
+     */
+    void SetWeight(
+        const int Index,
+        const ScalarType Value) override
+    {
+        auto& node = *Node(Index);
+
+        node.SetValue(Kratos::NURBS_CONTROLPOINT_WEIGHT, Value);
+    }
+
+    /** Gets the value of a nodal Kratos variable on a point at the curve.
+     * 
+     * \param Variable Kratos variable
+     * \param T Curve parameter
+     * 
+     * @return The value of the variable at the given curve point.
+     */
+    template <typename TDataType, typename TVariableType = Variable<TDataType>>
+    TDataType ValueAt(
+        const TVariableType& Variable,
+        const double T) const
+    {
+        return EvaluateAt<TDataType>([&](int i) -> TDataType {
+            return Node(i)->GetValue(Variable);
+        }, T);
+    }
+
+    /** Gets the derivatives of a nodal Kratos variable on a point at the curve.
+     * 
+     * \param Variable Kratos variable
+     * \param T Curve parameter
+     * \param Order Order of the highest derivative to compute
+     * 
+     * @return The value and the derivatives of the variable at the given
+     * curve point.
+     */
+    template <typename TDataType, typename TVariableType = Variable<TDataType>>
+    std::vector<TDataType> ValueAt(
+        const TVariableType& Variable,
+        const double T,
+        const int Order) const
+    {
+        return EvaluateAt<TDataType>([&](int i) -> TDataType {
```

Comment:
This is little bit obscure

Reply:
The code or the functionality?

Reply:
The code is OK, the think is that lambdas are not very extended in use. I recommend to add a description what is doing, to make easier to the people to understand in the future